### PR TITLE
Adds the ability to pass a `target` query parameter that is made avai…

### PR DESCRIPTION
**Pull request by stephen-soltesz**
Creating a PR so I can add comments in-line.

---
This change is [Reviewable](https://reviewable.io/reviews/nkinkade/script_exporter/1)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,6 +21,11 @@ scripts:
   - name: timeout
     script: sleep 5
     timeout: 1
+
+  - name: target
+    script: example.sh
+    timeout: 4
+    target: service.example.com
```
Comment (stephen-soltesz) [inline README.md]: While the configuration may make it possible to specify the `target` in a configuration file, I suggest leaving this part of the example out as it obscures the intended usage via environment variable.

Could the `script:` above correctly be shown as:
```
  - name: target
￼    script: example.sh ${TARGET}
    timeout: 4
```

File: README.md [hunk 2]
```diff
@@ -69,6 +74,14 @@ script_duration_seconds{script="success"} 5.013670
 script_success{script="success"} 1
 ```
 
+A `target` can be specified for each script in the configuration file, or, more
+usefully, passed as a parameter which is made available to the script(s) that
+gets run as an environment variable named `$TARGET`. This allows you to leverage
+Prometheus targets, if you happen to need the script to operate on an arbitrary
+number of remote hosts/services.
+
+`$ curl http://localhost:9172/probe?name=example.sh&target=service.example.com`
```
Comment (stephen-soltesz) [inline README.md]: Is the `name=` parameter correct based on the earlier example? I expected this to be `name=target`. Maybe a less ambiguous name here would be helpful. e.g. `name=script_with_parameter&target=service.example.com`.

File: README.md (hunk 1 above)
Comment (stephen-soltesz) [reply README.md]: Or is the expectation that `example.sh` inspect its environment before running?

File: script_exporter.go [hunk 3]
```diff
@@ -47,6 +48,7 @@ func runScript(script *Script) error {
 	defer cancel()
 
 	bashCmd := exec.CommandContext(ctx, *shell)
+	bashCmd.Env = append(os.Environ(), fmt.Sprintf("TARGET=%s", script.Target))
```
Comment (stephen-soltesz) [inline script_exporter.go:55]: What security concerns are there from an adversary injecting arbitrary text into the script's environment?

**Comment by nkinkade** [comment]
Thanks for looking at this, @stephen-soltesz. I've added some replies, but not yet any commits.

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[README.md, line 28 at r1](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-KyNCLEIeDfhCJ72GZ9n-r1-28:-KyQjaxwaB7DMqKKFSVf:bao2coo) ([raw file](https://github.com/nkinkade/script_exporter/blob/d3926976cd8aa16b5916c5534793e29d46462cde/README.md#L28)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Or is the expectation that `example.sh` inspect its environment before running?
</blockquote></details>

I think you're right about removing this example, as it's not a use case that anyone would find useful. If you're going to add a static `target` configuration in the YML config file in advance, then you may as  well just hard code the target in your script in the first place.

However, the usage you specify won't work, I think.  `$TARGET` won't be available in the YML config file, but actually to any code that gets run in the script itself. The reason I chose to go with an environment variable instead of passing the target as an argument to the script has to do with the way they implemented this.  They don't actually run the  script directly. Instead [they launch an interpreter (/bin/sh by default)](https://github.com/adhocteam/script_exporter/blob/master/script_exporter.go#L26), and then open a pipe to STDIN  of that process and then pipe in the contents of the script file you specify. This method didn't seem amenable to adding any arguments, so I chose to just add the target as an environment variable, which seemed good enough.

---

*[README.md, line 83 at r1](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-KyNCLEIeDfhCJ72GZ9n-r1-83:-KyQlR5BknTQoJFRU9On:b-e5nlye) ([raw file](https://github.com/nkinkade/script_exporter/blob/d3926976cd8aa16b5916c5534793e29d46462cde/README.md#L83)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Is the `name=` parameter correct based on the earlier example? I expected this to be `name=target`. Maybe a less ambiguous name here would be helpful. e.g. `name=script_with_parameter&target=service.example.com`.
</blockquote></details>

The `name=` parameter is one of their standard options. Granted, it's fairly generic and undescriptive. It just points to some particular YAML block, then it grabs the content of the `script` parameter to figure out what to actually run. I can give this a better name, because `example.sh` isn't very accurate here.

---

*[script_exporter.go, line 51 at r1](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-KyNCLEJ65PatF-3f0xz-r1-51:-KyQsp9FbOxi3msdL6YD:btsnyd) ([raw file](https://github.com/nkinkade/script_exporter/blob/d3926976cd8aa16b5916c5534793e29d46462cde/script_exporter.go#L51)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

What security concerns are there from an adversary injecting arbitrary text into the script's environment?
</blockquote></details>

There are large security considerations, if the person writing the script doesn't validate the input. I suppose I could have script_exporter match the input to some regex for a valid URL, since the only intended input should be a domain name or a URL. What do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-:-KyQtJggIdMKagabXZYS:b-vhnbxa)*

**Comment by nkinkade** [comment]
@stephen-soltesz: I've made some updates according to your feedback. PTAL?

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[README.md, line 28 at r1](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-KyNCLEIeDfhCJ72GZ9n-r1-28:-L1xuFZh13VJ_CWArURt:b8for0v) ([raw file](https://github.com/nkinkade/script_exporter/blob/d3926976cd8aa16b5916c5534793e29d46462cde/README.md#L28)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

I think you're right about removing this example, as it's not a use case that anyone would find useful. If you're going to add a static `target` configuration in the YML config file in advance, then you may as  well just hard code the target in your script in the first place.

However, the usage you specify won't work, I think.  `$TARGET` won't be available in the YML config file, but actually to any code that gets run in the script itself. The reason I chose to go with an environment variable instead of passing the target as an argument to the script has to do with the way they implemented this.  They don't actually run the  script directly. Instead [they launch an interpreter (/bin/sh by default)](https://github.com/adhocteam/script_exporter/blob/master/script_exporter.go#L26), and then open a pipe to STDIN  of that process and then pipe in the contents of the script file you specify. This method didn't seem amenable to adding any arguments, so I chose to just add the target as an environment variable, which seemed good enough.
</blockquote></details>

This should be fixed/improved now. I think.

---

*[script_exporter.go, line 51 at r1](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-KyNCLEJ65PatF-3f0xz-r1-51:-L1xu_AO5AZGmAorl8Eu:bmeys6q) ([raw file](https://github.com/nkinkade/script_exporter/blob/d3926976cd8aa16b5916c5534793e29d46462cde/script_exporter.go#L51)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

There are large security considerations, if the person writing the script doesn't validate the input. I suppose I could have script_exporter match the input to some regex for a valid URL, since the only intended input should be a domain name or a URL. What do you think?
</blockquote></details>

I now run the target value through a basic regex which essentially makes sure it doesn't contain anything but `/a-zA-Z0-9-./` and is not more than 253 characters. Does that seem sufficient to prevent malicious values being interpreted by the script?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-:-L1xuzLRoDbuJhMtKK28:by4obsk)*

**Comment by stephen-soltesz** [comment]
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[README.md, line 83 at r1](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-KyNCLEIeDfhCJ72GZ9n-r1-83:-L1yCTMK1WVjtEci0vMU:b-hlvk18) ([raw file](https://github.com/nkinkade/script_exporter/blob/d3926976cd8aa16b5916c5534793e29d46462cde/README.md#L83)):*
I suggest using the same name in the parameters here as defined in the sample configuration above. This helps the reader tie the sample above to the specific usage example here:

If the sample config is:

>   - name: target
>     script: example.sh ${TARGET}
>     timeout: 4

Then it would be called with parameters like:

```
name=target&target=service.example.com
```

My suggestions is that since `target` is also a very generic name, that a more descriptive example name be chosen in the sample configuration and here.

If the goal is for the upstream maintainer to adopt these changes, then clear docs are especially helpful to communicating with the author as well as the future users.

---

*[script_exporter.go, line 51 at r1](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-KyNCLEJ65PatF-3f0xz-r1-51:-L1yDX0wl5qfi3V0NNf8:b3lc67b) ([raw file](https://github.com/nkinkade/script_exporter/blob/d3926976cd8aa16b5916c5534793e29d46462cde/script_exporter.go#L51)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

I now run the target value through a basic regex which essentially makes sure it doesn't contain anything but `/a-zA-Z0-9-./` and is not more than 253 characters. Does that seem sufficient to prevent malicious values being interpreted by the script?
</blockquote></details>

Yes. That sounds good to me. A pedant could object that DNS now contains unicode characters, but I am not that person.

---

*[script_exporter.go, line 115 at r2](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-L1yFAksQHMcX6CyD2Ak:-L1yFAksQHMcX6CyD2Al:b7us385) ([raw file](https://github.com/nkinkade/script_exporter/blob/4360595c607e58db45e164a78abd36081bee4ef4/script_exporter.go#L115)):*
> ```Go
> 	var targetRegexp *regexp.Regexp
> 
> 	// A *very* basic regex pattern to be sure that the target looks minimally
> ```

I suggest removing "*very* basic" from comment.

Something simple and declarative like: "A regex pattern to match only ascii dns names to avoid special characters that could corrupt a shell environment."

I'm maybe being more picky only b/c you plan to merge this upstream.

---

*[script_exporter.go, line 117 at r2](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-L1yESN2nzPmPqIetm5q:-L1yESN2nzPmPqIetm5r:b1njvdv) ([raw file](https://github.com/nkinkade/script_exporter/blob/4360595c607e58db45e164a78abd36081bee4ef4/script_exporter.go#L117)):*
> ```Go
> 	// A *very* basic regex pattern to be sure that the target looks minimally
> 	// like a domain name and contains no special shell characters.
> 	var targetPattern string = "^[a-zA-Z0-9-.]{2,253}$"
> ```

I suggest making this a global constant using `regexp.MustCompile`.

For example after the flag defs at the start of the file:

```
var targetRegexp =  regexp.MustCompile("^[a-zA-Z0-9-.]{2,253}$")
```

---

*[script_exporter.go, line 129 at r2](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-L1yGNZl4-85KoWgto91:-L1yGNZl4-85KoWgto92:bjrpoar) ([raw file](https://github.com/nkinkade/script_exporter/blob/4360595c607e58db45e164a78abd36081bee4ef4/script_exporter.go#L129)):*
> ```Go
> 	targetRegexp, err = regexp.Compile(targetPattern)
> 	if target != "" && ! targetRegexp.MatchString(target) {
> 		log.Infof("The query parameter 'target' is not valid: %s", target)
> ```

Here `target` is the unmodified user-provided string. Is there any risk from dropping user-provided input into the log as-is?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nkinkade/script_exporter/1)*

**Comment by nkinkade** [comment]
@stephen-soltesz: I think I have addressed the comments you had. PTAL?

---

Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[script_exporter.go, line 115 at r2](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-L1yFAksQHMcX6CyD2Ak:-L1yMTjxX2Xkv7y0CHLH:bt2vr1b) ([raw file](https://github.com/nkinkade/script_exporter/blob/4360595c607e58db45e164a78abd36081bee4ef4/script_exporter.go#L115)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

I suggest removing "*very* basic" from comment.

Something simple and declarative like: "A regex pattern to match only ascii dns names to avoid special characters that could corrupt a shell environment."

I'm maybe being more picky only b/c you plan to merge this upstream.
</blockquote></details>

Fixed.

---

*[script_exporter.go, line 117 at r2](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-L1yESN2nzPmPqIetm5q:-L1yMWUTi-NBbGWDj-fA:b-896fix) ([raw file](https://github.com/nkinkade/script_exporter/blob/4360595c607e58db45e164a78abd36081bee4ef4/script_exporter.go#L117)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

I suggest making this a global constant using `regexp.MustCompile`.

For example after the flag defs at the start of the file:

```
var targetRegexp =  regexp.MustCompile("^[a-zA-Z0-9-.]{2,253}$")
```
</blockquote></details>

Done.

---

*[script_exporter.go, line 129 at r2](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-L1yGNZl4-85KoWgto91:-L1yMa4TpAMkox2AoN3s:b6ie76d) ([raw file](https://github.com/nkinkade/script_exporter/blob/4360595c607e58db45e164a78abd36081bee4ef4/script_exporter.go#L129)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Here `target` is the unmodified user-provided string. Is there any risk from dropping user-provided input into the log as-is?
</blockquote></details>

I suppose I don't even need to log this, but simply return.  They don't log when `pattern` fails to compile. I was logging here mostly as a way to debug, and figured I'd leave it  as potentially useful in the future in understanding if this feature was trying to be exploited in some way. I'm just going to remove the logging.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-:-L1yO7bMhtqiC0fKMqtN:b-86e831)*

**Comment by stephen-soltesz** [comment]
:lgtm:

---

Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[script_exporter.go, line 129 at r2](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-L1yGNZl4-85KoWgto91:-L1yPVx7GEJx9BDQ7H6b:b-81nlyc) ([raw file](https://github.com/nkinkade/script_exporter/blob/4360595c607e58db45e164a78abd36081bee4ef4/script_exporter.go#L129)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

I suppose I don't even need to log this, but simply return.  They don't log when `pattern` fails to compile. I was logging here mostly as a way to debug, and figured I'd leave it  as potentially useful in the future in understanding if this feature was trying to be exploited in some way. I'm just going to remove the logging.
</blockquote></details>

SGTM.

---

*[script_exporter.go, line 117 at r3](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-L1yPo_P7betMDN-k6Zt:-L1yP4RYlYcN4sc6ZgHx:b-4uunwo) ([raw file](https://github.com/nkinkade/script_exporter/blob/b10e99bf08dcd61b87c0735998a1e263b2d0dbea/script_exporter.go#L117)):*
> ```Go
> 	// prevent inadvertent or malicious injection of special shell characters
> 	// into the scripts environment.
> 	var targetRegexp = regexp.MustCompile("^[a-zA-Z0-9-.]{4,253}$")
> ```

I would make this a global constant since the value never changes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-:-L1yPlMhDwNeiI_Qp_p9:bnfp4nl)*

**Comment by nkinkade** [comment]
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[script_exporter.go, line 117 at r3](https://reviewable.io:443/reviews/nkinkade/script_exporter/1#-L1yPo_P7betMDN-k6Zt:-L1yVv5M5AyfMkAezptO:b-896fix) ([raw file](https://github.com/nkinkade/script_exporter/blob/b10e99bf08dcd61b87c0735998a1e263b2d0dbea/script_exporter.go#L117)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

I would make this a global constant since the value never changes.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nkinkade/script_exporter/1)*

**Pull request merged**